### PR TITLE
Support interfaces

### PR DESCRIPTION
**Pull request by spawnia**
- [x] Added automated tests
- [x] Documented for all relevant versions
- [x] Updated the changelog

Resolves https://github.com/spawnia/sailor/issues/35

**Changes**

- Support interfaces

**Breaking changes**

- Always add field `__typename` to any subselection and have it available in the result through `TypedObject`
- Name generated `TypedObject` classes after the corresponding object type

### CODE REVIEW

**Comment by spawnia** [comment]
@simPod even though I said I would not do it, your issue sparked a fire in me to see if interfaces can be supported easily. What do you think of this design from a client perspective?

**Comment by simPod** [comment]
Thanks for doing this!

I have read the examples and I don't understand how would I read `id` on any object implementing Node

```gql
query UserOrPost($id: ID!) {
    node(id: $id) {
        id
    }
}```

e.g. what would this query generate?

**Comment by spawnia** [comment]
> e.g. what would this query generate?

Stripped down versions of the generated `User` and `Post` classes with only a single property `$id`. That leads me to a good point to consider, we will have to generate code for all possible implementations of the interface. I will extend the example and hopefully make it more clear that way.

**Review by simPod — COMMENTED** [review]
I still would say we have need for interfaces.

Because I don't want to do this 

```php
assert($userOrPost instanceof \Spawnia\Sailor\Simple\UserOrPost\UserOrPost\User);
```

When I'm interested only in Node/Interface properties. From the query I know that the property `id` is always present on `$userOrPost`.

**Comment by spawnia** [comment]
> From the query I know that the property `id` is always present on `$userOrPost`.

And you will know it on the type level as well, given that every member of the union `Post|Task|User` has a property `$id`.

**Comment by simPod** [comment]
But then I'd have to rely on knowing what type implementations are there or to know at least one and then type against it.

I'd rather type against the same structure as is in my query. So when I don't deal with types in gql query but interfaces, I don't want to deal with types in code either. E.g. then when type I type against disappears, I would not have to change the query but would have to change the code.

------------

hm I wrote that but then checked I can do this so I guess it's fine as it is https://phpstan.org/r/1e5a6e05-85a4-4f90-bdb9-c80620e8b4d7 👌🏾

File: examples/polymorphic/expected/UserOrPost/UserOrPost.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Spawnia\Sailor\Polymorphic\UserOrPost;
+
+class UserOrPost extends \Spawnia\Sailor\TypedObject
+{
+    /** @var \Spawnia\Sailor\Polymorphic\UserOrPost\Node\User|\Spawnia\Sailor\Polymorphic\UserOrPost\Node\Post|\Spawnia\Sailor\Polymorphic\UserOrPost\Node\Task */
```
Comment (simPod) [inline examples/polymorphic/expected/UserOrPost/UserOrPost.php:9]: Not sure if doable but this could use import for better readability

**Review by simPod — APPROVED** [review]
Awesome

File: examples/polymorphic/expected/UserOrPost/UserOrPost.php (hunk 1 above)
Comment (spawnia) [reply examples/polymorphic/expected/UserOrPost/UserOrPost.php:9]: Agreed on the readability. `nette/php-generator` does not support adding imports, so I will hold off on investing extra work into a cosmetic change that may or may not improve the library.

**Pull request merged**